# Mini-project

## 1. Student  
Yassine TAHARASTE  
041029-T314  
1159492902    
yastah-5     
  
## 2. Targeted Grading Criteria
Grade 5 

In [1]:
import os
import numpy as np
import pandas as pd
import math
import time
from sklearn.preprocessing import StandardScaler, LabelEncoder
from sklearn.model_selection import train_test_split, StratifiedKFold, KFold
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier
from sklearn.svm import SVC
from sklearn.metrics import accuracy_score, f1_score, classification_report, cohen_kappa_score, silhouette_score, davies_bouldin_score
from sklearn.cluster import KMeans
from sklearn.mixture import GaussianMixture
from collections import Counter
from typing import Tuple
from scipy.io import arff
from IPython.display import display, Markdown


 ## 3. Datasets Used
  
  22 datasets  

**Iris**

* Number of samples: 150
* Number of features: 4 (continuous)
* Feature value range: approximately 0.1 to 7.9
* Number of classes: 3

**Wine**

* Number of samples: 178
* Number of features: 13 (continuous)
* Feature value range: varies across features
* Number of classes: 3

**Waveform**

* Number of samples: 5,000
* Number of features: 21 (continuous)
* Feature value range: continuous, synthetic data
* Number of classes: 3

**Car Evaluation**

* Number of samples: 1,728
* Number of features: 6 (categorical)
* Feature value range: discrete values
* Number of classes: 4

**Titanic**

* Number of samples: 2,201
* Number of features: 3–4 (categorical and numerical, depending on preprocessing)
* Feature value range: mixed
* Number of classes: 2

**Ionosphere**

* Number of samples: 351
* Number of features: 34 (continuous)
* Feature value range: approximately −1 to 1
* Number of classes: 2

**Dermatology**

* Number of samples: 366
* Number of features: 34 (mostly integer-valued)
* Feature value range: small integer ranges
* Number of classes: 6

**Ecoli**

* Number of samples: 336
* Number of features: 7 (continuous)
* Feature value range: mostly between 0 and 1
* Number of classes: 8

**Zoo**

* Number of samples: 101
* Number of features: 16 (binary and categorical)
* Feature value range: discrete
* Number of classes: 7

**Page Blocks**

* Number of samples: 5,473
* Number of features: 10 (numerical)
* Feature value range: varies by feature
* Number of classes: 5

**Spambase**

* Number of samples: 4,601
* Number of features: 57 (continuous)
* Feature value range: 0 to 100
* Number of classes: 2

**Acute Inflammation**

* Number of samples: 120
* Number of features: 6 (binary/categorical)
* Feature value range: discrete
* Number of classes: 2

**Vertebral Column (3 classes)**

* Number of samples: 310
* Number of features: 6 (continuous)
* Feature value range: varies across features
* Number of classes: 3

**Vertebral Column (2 classes)**

* Number of samples: 310
* Number of features: 6 (continuous)
* Feature value range: varies across features
* Number of classes: 2

**Tic-Tac-Toe**

* Number of samples: 958
* Number of features: 9 (categorical)
* Feature value range: discrete
* Number of classes: 2

**Synthetic Control**

* Number of samples: 600
* Number of features: 60 (continuous, time-series)
* Feature value range: continuous
* Number of classes: 6

**Statlog Vehicle**

* Number of samples: 846
* Number of features: 18 (numerical)
* Feature value range: varies across features
* Number of classes: 4

**Statlog Heart**

* Number of samples: 270
* Number of features: 13 (numerical and categorical)
* Feature value range: mixed
* Number of classes: 2

**Seeds**

* Number of samples: 210
* Number of features: 7 (continuous)
* Feature value range: varies across features
* Number of classes: 3

**Balance Scale**

* Number of samples: 625
* Number of features: 4 (discrete numerical)
* Feature value range: small integer values
* Number of classes: 3

**Acute Nephritis**

* Number of samples: 120
* Number of features: 6 (binary/categorical)
* Feature value range: discrete
* Number of classes: 2

**Heart (Hungarian)**

* Number of samples: 294
* Number of features: 13 (numerical and categorical)
* Feature value range: mixed
* Number of classes: 2

In [2]:
datasets = ['iris', 'wine', 'waveform', 'car', 'titanic', 'ionosphere', 
            'dermatology', 'ecoli', 'zoo', 'page-blocks', 'spambase', 
            'acute-inflammation', 'vertebral-column-3clases', 
            'vertebral-column-2clases', 'tic-tac-toe', 'synthetic-control', 
            'statlog-vehicle', 'statlog-heart', 'seeds', 'balance-scale', 
            'acute-nephritis', 'heart-hungarian']

## 4. Models Used  
  
Supervised :  
- `LogisticRegression`  
- `RandomForestClassifier`  
- `MLP` (handmade implementation)  
  
Unsupervised:  
- `KMeans`  
- `GaussianMixture` / GMM  
- `SOM` (handmade implementation)  

## 5. Experimental Methodology
  
Preprocessing
- Missing and infinite values are replace by mean values
- Non-numeric features and target labels are encoded with `LabelEncoder` 
- Features are standardized with `StandardScaler` fitted on the training partition and applied to validation/test to avoid leakage.
- Constant columns and candidate ID columns (columns with a unique value per row) are removed

Validation and Splits
- Each dataset is split into train+validation (80%) and test (20%).
- Hyperparameter selection is performed by Stratified K-Fold cross-validation on the train+validation set (`n_splits=5`).
- If `StratifiedKFold` cannot be applied due to extremely small class counts, a fallback strategy assigns rare-class samples to training folds to preserve class representation during validation.

Training and Evaluation
- For every CV fold, all preprocessing steps (imputation, encoding, scaling) are fit on the training fold and then applied to the corresponding validation fold to prevent data leakage.
- After hyperparameter tuning, the selected model is retrained on the complete train+validation set and evaluated once on the held-out test set.
- Primary reported metrics are `accuracy` and `f1_macro` averaged across folds. 
- For unsupervised methods, cluster labels are mapped to class labels by majority vote on the training data before computing supervised metrics. 
- I chose to tune the models using supervised metrics (accuracy and f1-score) rather than metrics that are more commonly used for unsupervised methods. I considered this choice more relevant because it allows a direct comparison between the results of these methods and those reported in the reference article. However, this decision is debatable. An alternative approach would have been to tune the models using intrinsic clustering metrics such as the silhouette score, Davies–Bouldin index, or Calinski–Harabasz score, which better reflect the internal structure and cohesion of the clusters independently of the ground-truth labels.
- For imbalanced datasets the selection criterion favors `f1_macro` over `accuracy` when class imbalance exceeds a predefined ratio.
- For final test, we also add the metric `cohen_kappa`. For unsupervised methods, clustering quality is reported using `silhouette_score` and `davies_bouldin_score`.

Reproducibility and Implementation Details
- Random seeds are fixed (`random_state=42`) for stochastic algorithms when supported to promote reproducibility.
- All fits (preprocessing and model training) use only training data; results are averaged across folds and saved to CSV for traceability.

## Load Datasets

In [ ]:
def load_dataset(name: str) -> Tuple[pd.DataFrame, pd.Series]:
    folder = os.path.join("data", name)
    files = [f for f in os.listdir(folder) if os.path.isfile(os.path.join(folder, f))]
    
    # pick data file
    arff_file = f"{name}.arff"
    data_file = f"{name}.data"

    if arff_file in files:
        chosen = arff_file
        is_arff = True
    elif data_file in files:
        chosen = data_file
        is_arff = False
    else:
        raise FileNotFoundError(f"Neither '{arff_file}' nor '{data_file}' found in folder {folder}")
    path = os.path.join(folder, chosen)
    
    if is_arff:
        raw_data, meta = arff.loadarff(path)
        df = pd.DataFrame(raw_data)
        # decode if needed
        for col in df.columns:
            if df[col].dtype == object:
                df[col] = df[col].apply(lambda x: x.decode('utf-8') if isinstance(x, bytes) else x)
    else:
        if name.lower().startswith("dermatology"): # exception
            df = pd.read_csv(path, header=None, sep=",", on_bad_lines='skip')
        else:
            df = pd.read_csv(path, header=None, sep=r"\s+", on_bad_lines='skip')

    X_raw = df.iloc[:, :-1]
    y = df.iloc[:, -1].astype(str).str.strip()
    
    # remove constant columns
    X_raw = X_raw.loc[:, X_raw.nunique() > 1]

    # column ID to remove if present
    for col in X_raw.columns:
        if X_raw[col].nunique() == len(X_raw):
            X_raw = X_raw.drop(columns=[col])
            break

    X = pd.DataFrame()
    for col in X_raw.columns:
        # try convertion
        converted = pd.to_numeric(X_raw[col], errors='coerce')
        if converted.notna().mean() > 0.8:
            # numeric
            X[col] = converted
        else:
            # encoding to 0,1,2,...
            le = LabelEncoder()
            X[col] = le.fit_transform(X_raw[col].astype(str))

    return X, y

# Verification
for d in datasets:
    print(f"\nDataset : {d}")
    X, y = load_dataset(d)
    print(X.head())
    print("Labels :", y.unique()[:10])



## Preprocessing

In [4]:
def preprocess(X_train: pd.DataFrame, X_test: pd.DataFrame):
    # - Impute missing values with column mean (fit on train), and apply the same fill to test
    # - fit StandardScaler on train.
    X_train = X_train.copy()
    X_test = X_test.copy()

    for col in X_train.columns:
        tr_col = pd.to_numeric(X_train[col].replace([np.inf, -np.inf], np.nan), errors='coerce')
        mean_val = tr_col.mean()
        X_train[col] = tr_col.fillna(mean_val)
        # apply same transformation to test (use train mean) because we must avoid data leakage
        X_test[col] = pd.to_numeric(X_test[col].replace([np.inf, -np.inf], np.nan), errors='coerce').fillna(mean_val)

    scaler = StandardScaler()
    # as before, we just fit scaler on train only to avoid data leakage
    scaler.fit(X_train.values)
    Xtr_scaled = pd.DataFrame(scaler.transform(X_train.values), columns=X_train.columns, index=X_train.index)
    Xt_scaled = pd.DataFrame(scaler.transform(X_test.values), columns=X_test.columns, index=X_test.index)
    return Xtr_scaled, Xt_scaled, scaler

def encode_labels(labels_train: pd.Series, labels_test: pd.Series):
    le = LabelEncoder()
    le.fit(labels_train)
    ytr = le.transform(labels_train)
    yt = le.transform(labels_test)
    return ytr, yt, le

def split_cv(X, y, n_splits=5, shuffle=True, random_state=42):
    try:
        skf = StratifiedKFold(n_splits=n_splits, shuffle=shuffle, random_state=random_state)
        folds = []
        for train_idx, test_idx in skf.split(X, y):
            folds.append((train_idx, test_idx))
        return folds

    except ValueError as e:
        print("StratifiedKFold failed:")
        pass

    #Fallback: isolate rare classes
    counts = Counter(y)
    rare_classes = {c for c, n in counts.items() if n < n_splits}

    rare_mask = np.array([c in rare_classes for c in y])
    common_mask = ~rare_mask

    idx = np.arange(len(y))
    rare_idx = idx[rare_mask]
    common_idx = idx[common_mask]

    # Stratified CV on common classes only
    skf_common = StratifiedKFold(n_splits=n_splits, shuffle=shuffle, random_state=random_state)

    folds = []
    for train_c, test_c in skf_common.split(np.zeros(len(common_idx)), y[common_mask]):
        #then we add the rare classes in the train set
        train_idx = np.concatenate([common_idx[train_c], rare_idx]) 
        test_idx = common_idx[test_c]
        folds.append((train_idx, test_idx))
    return folds

## Training and validation for supervised methods

In [5]:
def train_and_evaluate_supervised(model, X_train, y_train, X_test, y_test, validation = True):
    model.fit(X_train, y_train)
    ypred = model.predict(X_test)
    acc = accuracy_score(y_test, ypred)
    f1 = f1_score(y_test, ypred, average='macro')
    if not validation:
        kappa = cohen_kappa_score(y_test, ypred)
        return {'accuracy': acc, 'f1_macro': f1, 'cohen_kappa': kappa, 'y_pred': ypred}
    return {'accuracy': acc, 'f1_macro': f1, 'y_pred': ypred}

In [ ]:
#example with logistic regression
X, y = load_dataset("iris")
X_tv, X_test, y_tv, y_test = train_test_split(X, y, test_size=0.2, stratify=y, random_state=42)
C_values = [0.01, 0.1, 1, 10]
results = []
splits = split_cv(X_tv, y_tv, n_splits=5)

for C in C_values:
    for fold_idx, (train_idx, val_idx) in enumerate(splits):
        X_train = X_tv.iloc[train_idx]
        y_train = y_tv.iloc[train_idx]
        X_val = X_tv.iloc[val_idx]
        y_val = y_tv.iloc[val_idx]

        Xtr_s, Xval_s, scaler = preprocess(X_train, X_val)
        ytr_e, yval_e, le = encode_labels(y_train, y_val)

        model = LogisticRegression(C=C, max_iter=1000, random_state=42)
        scores = train_and_evaluate_supervised(model, Xtr_s, ytr_e, Xval_s, yval_e)

        results.append({
            "C": C,
            "fold": fold_idx,
            "accuracy": scores["accuracy"],
            "f1_macro": scores["f1_macro"]
        })

df_results = pd.DataFrame(results)
mean_scores = df_results.groupby("C").mean()
print(mean_scores)

counter = Counter(y_tv)
ratio = max(counter.values()) / min(counter.values())
if ratio > 2:
    metric_to_optimize = "f1_macro" # favor F1-score for imbalanced datasets
else:
    metric_to_optimize = "accuracy"
best_C = mean_scores[metric_to_optimize].idxmax()
print("Selected C:", best_C)

Xtr_s, Xtst_s, scaler = preprocess(X_tv, X_test)
ytr_e, ytst_e, le = encode_labels(y_tv, y_test)

model = LogisticRegression(C=best_C, max_iter=1000, random_state=42)
final_scores = train_and_evaluate_supervised(model, Xtr_s, ytr_e, Xtst_s, ytst_e, validation=False)
print("Performance on test:", final_scores)

       fold  accuracy  f1_macro
C                              
0.01    2.0  0.875000  0.872159
0.10    2.0  0.916667  0.915808
1.00    2.0  0.958333  0.957971
10.00   2.0  0.966667  0.966337
Selected C: 10.0
Performance on test: {'accuracy': 1.0, 'f1_macro': 1.0, 'cohen_kappa': 1.0, 'y_pred': array([0, 2, 1, 1, 0, 1, 0, 0, 2, 1, 2, 2, 2, 1, 0, 0, 0, 1, 1, 2, 0, 2,
       1, 2, 2, 1, 1, 0, 2, 0])}


In [7]:
def train_and_evaluate_unsupervised(model, X_train, y_train, X_test, y_test, validation = True):
    model.fit(X_train)

    #assign class to clusters
    cluster_to_class = {}
    clusters_train = model.predict(X_train)
    for c in np.unique(clusters_train):
        indices = np.where(clusters_train == c)[0]
        majority_class = Counter(y_train[indices]).most_common(1)[0][0]
        cluster_to_class[c] = majority_class

    clusters_test = model.predict(X_test)
    y_pred = np.array([cluster_to_class.get(c, -1) for c in clusters_test])
    acc = accuracy_score(y_test, y_pred)
    f1 = f1_score(y_test, y_pred, average='macro')

    if not validation:
        kappa = cohen_kappa_score(y_test, y_pred)
        labels = model.predict(X_train)
        silhouette = silhouette_score(X_train, labels)
        davies_bouldin = davies_bouldin_score(X_train, labels)
        return {'accuracy': acc, 'f1_macro': f1, 'cohen_kappa': kappa, 'silhouette': silhouette, 'davies_bouldin': davies_bouldin, 'y_pred': y_pred}

    return {'accuracy': acc, 'f1_macro': f1, 'y_pred': y_pred}

In [ ]:
# example with GMM
X, y = load_dataset("iris")
X_tv, X_test, y_tv, y_test = train_test_split(X, y, test_size=0.2, stratify=y, random_state=42)
splits = split_cv(X_tv, y_tv, n_splits=5)
n_components_values = [2, 3, 4, 5] # hyperparameter
results = []

for n_components in n_components_values:
    for fold_idx, (train_idx, val_idx) in enumerate(splits):
        X_train = X_tv.iloc[train_idx]
        y_train = y_tv.iloc[train_idx]
        X_val = X_tv.iloc[val_idx]
        y_val = y_tv.iloc[val_idx]

        Xtr_s, Xval_s, scaler = preprocess(X_train, X_val)
        ytr_e, yval_e, le = encode_labels(y_train, y_val)

        model = GaussianMixture(n_components=n_components, random_state=42)
        scores = train_and_evaluate_unsupervised(model, Xtr_s, ytr_e, Xval_s, yval_e)
        results.append({
            "n_components": n_components,
            "fold": fold_idx,
            "accuracy": scores["accuracy"],
            "f1_macro": scores["f1_macro"]
        })

df_results = pd.DataFrame(results)
mean_scores = df_results.groupby("n_components").mean()
print(mean_scores)

counter = Counter(y_tv)
ratio = max(counter.values()) / min(counter.values())
if ratio > 1.5:
    metric_to_optimize = "f1_macro" # favor F1-score for imbalanced datasets
else:
    metric_to_optimize = "accuracy"
best_n_components = mean_scores[metric_to_optimize].idxmax()
print("Selected n_components:", best_n_components)

Xtr_s, Xtst_s, scaler = preprocess(X_tv, X_test)
ytr_e, ytst_e, le = encode_labels(y_tv, y_test)

model = GaussianMixture(n_components=best_n_components, random_state=42)
final_scores = train_and_evaluate_unsupervised(model, Xtr_s, ytr_e, Xtst_s, ytst_e, False)
print("Performance on test:", final_scores)


              fold  accuracy  f1_macro
n_components                          
2              2.0  0.666667  0.555556
3              2.0  0.900000  0.894315
4              2.0  0.875000  0.867325
5              2.0  0.866667  0.857680
Selected n_components: 3
Performance on test: {'accuracy': 1.0, 'f1_macro': 1.0, 'cohen_kappa': 1.0, 'silhouette': 0.3879555851415599, 'davies_bouldin': 1.0386632303083412, 'y_pred': array([0, 2, 1, 1, 0, 1, 0, 0, 2, 1, 2, 2, 2, 1, 0, 0, 0, 1, 1, 2, 0, 2,
       1, 2, 2, 1, 1, 0, 2, 0])}


## Models Implemented
### SOM (Unsupervised)

In [9]:
def getEuclideanDistance(single_point, array):
    nrows, ncols, nfeatures=array.shape[0],array.shape[1], array.shape[2]
    points = array.reshape((nrows * ncols, nfeatures))
    dist = np.sqrt(np.sum((points - single_point) ** 2, axis=1))
    return dist.reshape((nrows, ncols))


class SOM:
    def __init__(self, dispRes, trainingData, ndim=10, nepochs=10, eta0=0.1, etadecay=0.05, sgm0=20, sgmdecay=0.05, showMode=0):
        self.dispRes = dispRes
        self.trainingData = trainingData
        self.ndim = ndim
        self.nepochs = nepochs
        self.eta0 = eta0
        self.etadecay = etadecay
        self.sgm0 = sgm0
        self.sgmdecay = sgmdecay
        self.showMode = showMode
        self.nfeatures = trainingData.shape[1]
        self.ntrainingvectors = trainingData.shape[0]
        self.nrows = ndim
        self.ncols = ndim
        self.som = None


    def train(self):
        mu, sigma = 0, 0.1
        np.random.seed(int(time.time()))
        self.som = np.random.normal(mu, sigma, (self.nrows, self.ncols, self.nfeatures))

        x, y = np.meshgrid(range(self.ncols), range(self.nrows))

        for t in range(1, self.nepochs+1):
            eta = self.eta0 * math.exp(-t*self.etadecay)
            sgm = self.sgm0 * math.exp(-t*self.sgmdecay)
            width = math.ceil(sgm*3)

            for ntraining in range(self.ntrainingvectors):
                trainingVector = self.trainingData[ntraining,:]
                dist = getEuclideanDistance(trainingVector, self.som)
                bmurow, bmucol = np.unravel_index(np.argmin(dist, axis=None), dist.shape)
                g = np.exp(-((np.power(x - bmucol,2)) + (np.power(y - bmurow,2))) / (2*sgm*sgm))
                fromrow = max(0,bmurow - width)
                torow   = min(bmurow + width,self.nrows)
                fromcol = max(0,bmucol - width)
                tocol   = min(bmucol + width,self.ncols)
                neighbourNeurons = self.som[fromrow:torow,fromcol:tocol,:]
                sz = neighbourNeurons.shape
                T = np.tile(trainingVector, (sz[0]*sz[1],1)).reshape((sz[0], sz[1], self.nfeatures))
                G = np.dstack([g[fromrow:torow,fromcol:tocol]]*self.nfeatures)
                neighbourNeurons = neighbourNeurons + eta * G * (T - neighbourNeurons)
                self.som[fromrow:torow,fromcol:tocol,:] = neighbourNeurons

        return self.som

    def test(self, classes, grid_):
        nclasses = np.max(classes)
        som_cl = np.zeros((self.ndim, self.ndim, nclasses+1))

        # counting classes for each neuron
        for ntraining in range(self.ntrainingvectors):
            trainingVector = self.trainingData[ntraining,:]
            class_of_sample = classes[ntraining]
            dist = getEuclideanDistance(trainingVector, self.som)
            bmurow, bmucol = np.unravel_index(np.argmin(dist, axis=None), dist.shape)
            som_cl[bmurow, bmucol, class_of_sample] += 1

        # then majority class assignment
        for i in range(self.nrows):
            for j in range(self.ncols):
                grid_[i,j] = np.argmax(som_cl[i,j,:])

        return grid_

### MLP (Supervised)

In [10]:
#functions of non-linear activations
def f_sigmoid(X, deriv=False):
    if not deriv:
        return 1 / (1 + np.exp(-X))
    else:
        return f_sigmoid(X)*(1 - f_sigmoid(X))


def f_softmax(X):
    Z = np.sum(np.exp(X), axis=1)
    Z = Z.reshape(Z.shape[0], 1)
    return np.exp(X) / Z

def exit_with_err(err_str):
    print >> sys.stderr, err_str
    sys.exit(1)

#Functionality of a single hidden layer
class Layer:
    def __init__(self, size, batch_size, is_input=False, is_output=False,
                 activation=f_sigmoid):
        self.is_input = is_input
        self.is_output = is_output

        # Z is the matrix that holds output values
        self.Z = np.zeros((batch_size, size[0]))
        # The activation function is an externally defined function (with a
        # derivative) that is stored here
        self.activation = activation

        # W is the outgoing weight matrix for this layer
        self.W = None
        # S is the matrix that holds the inputs to this layer
        self.S = None
        # D is the matrix that holds the deltas for this layer
        self.D = None
        # Fp is the matrix that holds the derivatives of the activation function
        self.Fp = None

        if not is_input:
            self.S = np.zeros((batch_size, size[0]))
            self.D = np.zeros((batch_size, size[0]))

        if not is_output:
            self.W = np.random.normal(size=size, scale=1E-4)

        if not is_input and not is_output:
            self.Fp = np.zeros((size[0], batch_size))

    def forward_propagate(self):
        if self.is_input:
            return self.Z.dot(self.W)

        self.Z = self.activation(self.S)
        if self.is_output:
            return self.Z
        else:
            # For hidden layers, we add the bias values here
            self.Z = np.append(self.Z, np.ones((self.Z.shape[0], 1)), axis=1)
            self.Fp = self.activation(self.S, deriv=True).T
            return self.Z.dot(self.W)

class MultiLayerPerceptron:
    def __init__(self, layer_config, batch_size=100):
        self.layers = []
        self.num_layers = len(layer_config)
        self.minibatch_size = batch_size

        for i in range(self.num_layers-1):
            if i == 0:
                print ("Initializing input layer with size {0}.".format(layer_config[i]))
                # Here, we add an additional unit at the input for the bias
                # weight.
                self.layers.append(Layer([layer_config[i]+1, layer_config[i+1]],
                                         batch_size,
                                         is_input=True))
            else:
                print ("Initializing hidden layer with size {0}.".format(layer_config[i]))
                # Here we add an additional unit in the hidden layers for the
                # bias weight.
                self.layers.append(Layer([layer_config[i]+1, layer_config[i+1]],
                                         batch_size,
                                         activation=f_sigmoid))

        print ("Initializing output layer with size {0}.".format(layer_config[-1]))
        self.layers.append(Layer([layer_config[-1], None],
                                 batch_size,
                                 is_output=True,
                                 activation=f_softmax))
        print ("Done!")

    def forward_propagate(self, data):
        # We need to be sure to add bias values to the input
        self.layers[0].Z = np.append(data, np.ones((data.shape[0], 1)), axis=1)

        for i in range(self.num_layers-1):
            self.layers[i+1].S = self.layers[i].forward_propagate()
        return self.layers[-1].forward_propagate()

    def backpropagate(self, yhat, labels):
        self.layers[-1].D = (yhat - labels).T
        for i in range(self.num_layers-2, 0, -1):
            # We do not calculate deltas for the bias values
            W_nobias = self.layers[i].W[0:-1, :]
            
            self.layers[i].D = W_nobias.dot(self.layers[i+1].D) * self.layers[i].Fp

    def update_weights(self, eta):
        for i in range(0, self.num_layers-1):
            W_grad = -eta*(self.layers[i+1].D.dot(self.layers[i].Z)).T
            self.layers[i].W += W_grad

    def evaluate(self, train_data, train_labels, test_data, test_labels,
                 num_epochs=70, eta=0.05, eval_train=False, eval_test=True):

        N_train = len(train_labels)*len(train_labels[0])
        N_test = len(test_labels)*len(test_labels[0])

        print ("Training for {0} epochs...".format(num_epochs))
        for t in range(0, num_epochs):
            out_str = "[{0:4d}] ".format(t)

            for b_data, b_labels in zip(train_data, train_labels):
                output = self.forward_propagate(b_data)
                self.backpropagate(output, b_labels)

                self.update_weights(eta=eta)

            if eval_train:
                errs = 0
                for b_data, b_labels in zip(train_data, train_labels):
                    output = self.forward_propagate(b_data)
                    yhat = np.argmax(output, axis=1)
                    errs += np.sum(1-b_labels[np.arange(len(b_labels)), yhat])

                out_str = ("{0} Training error: {1:.5f}".format(out_str,
                                                           float(errs)/N_train))

            if eval_test:
                errs = 0
                for b_data, b_labels in zip(test_data, test_labels):
                    output = self.forward_propagate(b_data)
                    yhat = np.argmax(output, axis=1)
                    errs += np.sum(1-b_labels[np.arange(len(b_labels)), yhat])

                out_str = ("{0} Test error: {1:.5f}").format(out_str,
                                                       float(errs)/N_test)

            print (out_str)
def label_to_bit_vector(labels, nbits):
    bit_vector = np.zeros((labels.shape[0], nbits))
    for i in range(labels.shape[0]):
        bit_vector[i, labels[i]] = 1.0

    return bit_vector

def create_batches(data, labels, batch_size, nb_labels, create_bit_vector=False):
    N = data.shape[0]
    print ("Batch size {0}, the number of examples {1}.".format(batch_size,N))

    chunked_data = []
    chunked_labels = []
    idx = 0
    while idx < N:
        chunked_data.append(data[idx:idx+batch_size, :])
        if not create_bit_vector:
            chunked_labels.append(labels[idx:idx+batch_size])
        else:
            bit_vector = label_to_bit_vector(labels[idx:idx+batch_size], nb_labels)
            chunked_labels.append(bit_vector)

        idx += batch_size

    return chunked_data, chunked_labels

def prepare_for_backprop(batch_size, Train_images, Train_labels, Valid_images, Valid_labels):
    
    print ("Creating data...")
    all_labels = np.concatenate([Train_labels, Valid_labels])
    nb_labels = len(np.unique(all_labels))
    batched_train_data, batched_train_labels = create_batches(Train_images, Train_labels,
                                              batch_size,
                                              nb_labels,
                                              create_bit_vector=True)
    batched_valid_data, batched_valid_labels = create_batches(Valid_images, Valid_labels,
                                              batch_size,
                                              nb_labels,
                                              create_bit_vector=True)
    print ("Done!")


    return batched_train_data, batched_train_labels,  batched_valid_data, batched_valid_labels



## Train validation et test of the 6 models overs the 22 datasets

## Hyperparameter Tuning Functions

In [11]:
def tune_logistic_regression(X_tv, y_tv, X_test, y_test):
    C_values = [0.01, 0.1, 1, 10]
    results = []
    splits = split_cv(X_tv, y_tv, n_splits=5)
    
    for C in C_values:
        for fold_idx, (train_idx, val_idx) in enumerate(splits):
            X_train = X_tv.iloc[train_idx]
            y_train = y_tv.iloc[train_idx]
            X_val = X_tv.iloc[val_idx]
            y_val = y_tv.iloc[val_idx]
            
            Xtr_s, Xval_s, scaler = preprocess(X_train, X_val)
            ytr_e, yval_e, le = encode_labels(y_train, y_val)
            
            model = LogisticRegression(C=C, max_iter=1000, random_state=42)
            scores = train_and_evaluate_supervised(model, Xtr_s, ytr_e, Xval_s, yval_e)
            
            results.append({
                "C": C,
                "fold": fold_idx,
                "accuracy": scores["accuracy"],
                "f1_macro": scores["f1_macro"]
            })
    
    df_results = pd.DataFrame(results)
    mean_scores = df_results.groupby("C").mean()
    
    counter = Counter(y_tv)
    ratio = max(counter.values()) / min(counter.values())
    metric_to_optimize = "f1_macro" if ratio > 2 else "accuracy"
    best_C = mean_scores[metric_to_optimize].idxmax()

    Xtr_s, Xtst_s, scaler = preprocess(X_tv, X_test)
    ytr_e, ytst_e, le = encode_labels(y_tv, y_test)
    
    model = LogisticRegression(C=best_C, max_iter=1000, random_state=42)
    final_scores = train_and_evaluate_supervised(model, Xtr_s, ytr_e, Xtst_s, ytst_e, validation=False)
    
    return final_scores, best_C


def tune_random_forest(X_tv, y_tv, X_test, y_test):
    n_estimators_values = [100, 300]
    max_depth_values = [None, 10]
    results = []
    splits = split_cv(X_tv, y_tv, n_splits=5)
    
    for n_est in n_estimators_values:
        for max_d in max_depth_values:
            for fold_idx, (train_idx, val_idx) in enumerate(splits):
                X_train = X_tv.iloc[train_idx]
                y_train = y_tv.iloc[train_idx]
                X_val = X_tv.iloc[val_idx]
                y_val = y_tv.iloc[val_idx]
                
                Xtr_s, Xval_s, scaler = preprocess(X_train, X_val)
                ytr_e, yval_e, le = encode_labels(y_train, y_val)
                
                model = RandomForestClassifier(n_estimators=n_est, max_depth=max_d, random_state=42)
                scores = train_and_evaluate_supervised(model, Xtr_s, ytr_e, Xval_s, yval_e)
                
                results.append({
                    "n_estimators": n_est,
                    "max_depth": max_d,
                    "fold": fold_idx,
                    "accuracy": scores["accuracy"],
                    "f1_macro": scores["f1_macro"]
                })
    
    df_results = pd.DataFrame(results)
    mean_scores = df_results.groupby(["n_estimators", "max_depth"]).mean()
    
    counter = Counter(y_tv)
    ratio = max(counter.values()) / min(counter.values())
    metric_to_optimize = "f1_macro" if ratio > 2 else "accuracy"
    best_params = mean_scores[metric_to_optimize].idxmax()
    best_n_est, best_max_d = best_params

    best_n_est = int(best_n_est)
    best_max_d = None if pd.isna(best_max_d) else int(best_max_d)
    
    # Train final model
    Xtr_s, Xtst_s, scaler = preprocess(X_tv, X_test)
    ytr_e, ytst_e, le = encode_labels(y_tv, y_test)
    
    model = RandomForestClassifier(n_estimators=best_n_est, max_depth=best_max_d, random_state=42)
    final_scores = train_and_evaluate_supervised(model, Xtr_s, ytr_e, Xtst_s, ytst_e, validation=False)
    
    return final_scores, (best_n_est, best_max_d)


def tune_mlp(X_tv, y_tv, X_test, y_test):
    hidden_layers_values = [50, 100]
    learning_rate_values = [0.001, 0.01]
    results = []
    splits = split_cv(X_tv, y_tv, n_splits=5)
    
    for hidden_size in hidden_layers_values:
        for lr in learning_rate_values:
            for fold_idx, (train_idx, val_idx) in enumerate(splits):
                X_train = X_tv.iloc[train_idx]
                y_train = y_tv.iloc[train_idx]
                X_val = X_tv.iloc[val_idx]
                y_val = y_tv.iloc[val_idx]
                
                Xtr_s, Xval_s, scaler = preprocess(X_train, X_val)
                ytr_e, yval_e, le = encode_labels(y_train, y_val)
                
                # MLP
                n_classes = len(np.unique(ytr_e))
                n_features = Xtr_s.shape[1]
                n_train = Xtr_s.shape[0]
                batch_size = max([b for b in range(1, 33) if n_train % b == 0])
                
                bt_train, bl_train, bt_val, bl_val = prepare_for_backprop(
                    batch_size, Xtr_s.values, ytr_e, Xval_s.values, yval_e
                )
                
                layer_config = [n_features, hidden_size, n_classes]
                mlp = MultiLayerPerceptron(layer_config, batch_size=batch_size)
                mlp.evaluate(bt_train, bl_train, bt_val, bl_val, num_epochs=50,
                           eta=lr, eval_train=False, eval_test=False)
                
                outputs = mlp.forward_propagate(Xval_s.values)
                y_pred = np.argmax(outputs, axis=1)
                acc = accuracy_score(yval_e, y_pred)
                f1 = f1_score(yval_e, y_pred, average='macro')
                
                results.append({
                    "hidden_size": hidden_size,
                    "learning_rate": lr,
                    "fold": fold_idx,
                    "accuracy": acc,
                    "f1_macro": f1
                })
    
    df_results = pd.DataFrame(results)
    mean_scores = df_results.groupby(["hidden_size", "learning_rate"]).mean()
    
    counter = Counter(y_tv)
    ratio = max(counter.values()) / min(counter.values())
    metric_to_optimize = "f1_macro" if ratio > 2 else "accuracy"
    best_params = mean_scores[metric_to_optimize].idxmax()
    best_hidden, best_lr = best_params
    
    # Train final model
    Xtr_s, Xtst_s, scaler = preprocess(X_tv, X_test)
    ytr_e, ytst_e, le = encode_labels(y_tv, y_test)
    
    n_classes = len(np.unique(ytr_e))
    n_features = Xtr_s.shape[1]
    n_train = Xtr_s.shape[0]
    batch_size = max([b for b in range(1, 33) if n_train % b == 0])
    
    bt_train, bl_train, bt_test, bl_test = prepare_for_backprop(
        batch_size, Xtr_s.values, ytr_e, Xtst_s.values, ytst_e
    )
    
    layer_config = [n_features, best_hidden, n_classes]
    mlp = MultiLayerPerceptron(layer_config, batch_size=batch_size)
    mlp.evaluate(bt_train, bl_train, bt_test, bl_test, num_epochs=50,
               eta=best_lr, eval_train=False, eval_test=False)
    
    outputs = mlp.forward_propagate(Xtst_s.values)
    y_pred = np.argmax(outputs, axis=1)
    acc = accuracy_score(ytst_e, y_pred)
    f1 = f1_score(ytst_e, y_pred, average='macro')
    kappa = cohen_kappa_score(ytst_e, y_pred)
    
    final_scores = {'accuracy': acc, 'f1_macro': f1, 'cohen_kappa': kappa, 'y_pred': y_pred}
    
    return final_scores, (best_hidden, best_lr)


def tune_kmeans(X_tv, y_tv, X_test, y_test):
    n_clusters_values = [2, 3, 4]
    results = []
    splits = split_cv(X_tv, y_tv, n_splits=5)
    
    for n_clust in n_clusters_values:
        for fold_idx, (train_idx, val_idx) in enumerate(splits):
            X_train = X_tv.iloc[train_idx]
            y_train = y_tv.iloc[train_idx]
            X_val = X_tv.iloc[val_idx]
            y_val = y_tv.iloc[val_idx]
            
            Xtr_s, Xval_s, scaler = preprocess(X_train, X_val)
            ytr_e, yval_e, le = encode_labels(y_train, y_val)
            
            model = KMeans(n_clusters=n_clust, random_state=42, n_init=10)
            scores = train_and_evaluate_unsupervised(model, Xtr_s, ytr_e, Xval_s, yval_e)
            
            results.append({
                "n_clusters": n_clust,
                "fold": fold_idx,
                "accuracy": scores["accuracy"],
                "f1_macro": scores["f1_macro"]
            })
    
    df_results = pd.DataFrame(results)
    mean_scores = df_results.groupby("n_clusters").mean()
    
    counter = Counter(y_tv)
    ratio = max(counter.values()) / min(counter.values())
    metric_to_optimize = "f1_macro" if ratio > 1.5 else "accuracy"
    best_n_clust = mean_scores[metric_to_optimize].idxmax()
    
    # Train final model
    Xtr_s, Xtst_s, scaler = preprocess(X_tv, X_test)
    ytr_e, ytst_e, le = encode_labels(y_tv, y_test)
    
    model = KMeans(n_clusters=best_n_clust, random_state=42, n_init=10)
    final_scores = train_and_evaluate_unsupervised(model, Xtr_s, ytr_e, Xtst_s, ytst_e, validation=False)
    
    return final_scores, best_n_clust


def tune_gmm(X_tv, y_tv, X_test, y_test):
    n_components_values = [2, 3, 4, 5]
    results = []
    splits = split_cv(X_tv, y_tv, n_splits=5)
    
    for n_comp in n_components_values:
        for fold_idx, (train_idx, val_idx) in enumerate(splits):
            X_train = X_tv.iloc[train_idx]
            y_train = y_tv.iloc[train_idx]
            X_val = X_tv.iloc[val_idx]
            y_val = y_tv.iloc[val_idx]
            
            Xtr_s, Xval_s, scaler = preprocess(X_train, X_val)
            ytr_e, yval_e, le = encode_labels(y_train, y_val)
            
            model = GaussianMixture(n_components=n_comp, random_state=42)
            scores = train_and_evaluate_unsupervised(model, Xtr_s, ytr_e, Xval_s, yval_e)
            
            results.append({
                "n_components": n_comp,
                "fold": fold_idx,
                "accuracy": scores["accuracy"],
                "f1_macro": scores["f1_macro"]
            })
    
    df_results = pd.DataFrame(results)
    mean_scores = df_results.groupby("n_components").mean()
    
    counter = Counter(y_tv)
    ratio = max(counter.values()) / min(counter.values())
    metric_to_optimize = "f1_macro" if ratio > 1.5 else "accuracy"
    best_n_comp = mean_scores[metric_to_optimize].idxmax()
    
    # Train final model
    Xtr_s, Xtst_s, scaler = preprocess(X_tv, X_test)
    ytr_e, ytst_e, le = encode_labels(y_tv, y_test)
    
    model = GaussianMixture(n_components=best_n_comp, random_state=42)
    final_scores = train_and_evaluate_unsupervised(model, Xtr_s, ytr_e, Xtst_s, ytst_e, validation=False)
    
    return final_scores, best_n_comp


def tune_som(X_tv, y_tv, X_test, y_test):
    grid_values = [5, 10]
    learning_rate_values = [0.1, 0.5]
    results = []
    splits = split_cv(X_tv, y_tv, n_splits=5)
    
    for grid_size in grid_values:
        for eta in learning_rate_values:
            for fold_idx, (train_idx, val_idx) in enumerate(splits):
                X_train = X_tv.iloc[train_idx]
                y_train = y_tv.iloc[train_idx]
                X_val = X_tv.iloc[val_idx]
                y_val = y_tv.iloc[val_idx]
                
                Xtr_s, Xval_s, scaler = preprocess(X_train, X_val)
                ytr_e, yval_e, le = encode_labels(y_train, y_val)
                
                # Train SOM
                som = SOM(dispRes=None, trainingData=Xtr_s.values, ndim=grid_size, 
                         nepochs=10, eta0=eta)
                som.train()
                grid = np.zeros((som.nrows, som.ncols), dtype=int)
                som.test(ytr_e, grid)
                
                # Predict on validation
                def som_predict(som_obj, Xdf, grid_map):
                    preds = []
                    for p in Xdf.values:
                        dist = getEuclideanDistance(p, som_obj.som)
                        r, c = np.unravel_index(np.argmin(dist), dist.shape)
                        preds.append(int(grid_map[r, c]))
                    return np.array(preds)
                
                y_pred = som_predict(som, Xval_s, grid)
                acc = accuracy_score(yval_e, y_pred)
                f1 = f1_score(yval_e, y_pred, average='macro')
                
                results.append({
                    "grid_size": grid_size,
                    "learning_rate": eta,
                    "fold": fold_idx,
                    "accuracy": acc,
                    "f1_macro": f1
                })
    
    df_results = pd.DataFrame(results)
    mean_scores = df_results.groupby(["grid_size", "learning_rate"]).mean()
    
    counter = Counter(y_tv)
    ratio = max(counter.values()) / min(counter.values())
    metric_to_optimize = "f1_macro" if ratio > 1.5 else "accuracy"
    best_params = mean_scores[metric_to_optimize].idxmax()
    best_grid, best_eta = best_params
    
    # Train final model
    Xtr_s, Xtst_s, scaler = preprocess(X_tv, X_test)
    ytr_e, ytst_e, le = encode_labels(y_tv, y_test)
    
    som = SOM(dispRes=None, trainingData=Xtr_s.values, ndim=best_grid, 
             nepochs=10, eta0=best_eta)
    som.train()
    grid = np.zeros((som.nrows, som.ncols), dtype=int)
    som.test(ytr_e, grid)
    
    def som_predict(som_obj, Xdf, grid_map):
        preds = []
        for p in Xdf.values:
            dist = getEuclideanDistance(p, som_obj.som)
            r, c = np.unravel_index(np.argmin(dist), dist.shape)
            preds.append(int(grid_map[r, c]))
        return np.array(preds)
    
    y_pred = som_predict(som, Xtst_s, grid)
    acc = accuracy_score(ytst_e, y_pred)
    f1 = f1_score(ytst_e, y_pred, average='macro')
    kappa = cohen_kappa_score(ytst_e, y_pred)
    
    # Compute clustering metrics on training data
    train_bmu = []
    for p in Xtr_s.values:
        dist = getEuclideanDistance(p, som.som)
        r, c = np.unravel_index(np.argmin(dist), dist.shape)
        train_bmu.append(r * som.ncols + c)
    train_bmu = np.array(train_bmu)
    
    silhouette = silhouette_score(Xtr_s.values, train_bmu)
    davies_bouldin = davies_bouldin_score(Xtr_s.values, train_bmu)
    
    final_scores = {'accuracy': acc, 'f1_macro': f1, 'cohen_kappa': kappa, 
                   'silhouette': silhouette, 'davies_bouldin': davies_bouldin, 'y_pred': y_pred}
    
    return final_scores, (best_grid, best_eta)

## Main Evaluation Loop with Hyperparameter Tuning

In [ ]:
EVAL_DATASETS = ['iris', 'wine', 'waveform', 'car', 'titanic', 'ionosphere', 
                 'dermatology', 'ecoli', 'zoo', 'page-blocks', 'spambase', 
                 'acute-inflammation', 'vertebral-column-3clases', 
                 'vertebral-column-2clases', 'tic-tac-toe', 'synthetic-control', 
                 'statlog-vehicle', 'statlog-heart', 'seeds', 'balance-scale', 
                 'acute-nephritis', 'heart-hungarian']
all_results = []
best_hyperparams = []

for ds_name in EVAL_DATASETS:
    print(f"dataset: {ds_name}")
    try:
        X, y = load_dataset(ds_name)
        # Split train+validation (80%) and test (20%)
        X_tv, X_test, y_tv, y_test = train_test_split(X, y, test_size=0.2, 
                                                      stratify=y, 
                                                      random_state=42)
        
        # LogisticRegression
        print("\nTuning LogisticRegression...")
        lr_scores, lr_params = tune_logistic_regression(X_tv, y_tv, X_test, y_test)
        all_results.append({
            'dataset': ds_name,
            'model': 'LogisticRegression',
            'accuracy': lr_scores['accuracy'],
            'f1_macro': lr_scores['f1_macro'],
            'cohen_kappa': lr_scores['cohen_kappa']
        })
        best_hyperparams.append({
            'dataset': ds_name,
            'model': 'LogisticRegression',
            'hyperparams': f"C={lr_params}"
        })
        
        # RandomForest
        print("Tuning RandomForest...")
        rf_scores, rf_params = tune_random_forest(X_tv, y_tv, X_test, y_test)
        all_results.append({
            'dataset': ds_name,
            'model': 'RandomForest',
            'accuracy': rf_scores['accuracy'],
            'f1_macro': rf_scores['f1_macro'],
            'cohen_kappa': rf_scores['cohen_kappa']
        })
        best_hyperparams.append({
            'dataset': ds_name,
            'model': 'RandomForest',
            'hyperparams': f"n_estimators={rf_params[0]}, max_depth={rf_params[1]}"
        })
        
        # MLP
        print("Tuning MLP...")
        mlp_scores, mlp_params = tune_mlp(X_tv, y_tv, X_test, y_test)
        all_results.append({
            'dataset': ds_name,
            'model': 'MLP',
            'accuracy': mlp_scores['accuracy'],
            'f1_macro': mlp_scores['f1_macro'],
            'cohen_kappa': mlp_scores['cohen_kappa']
        })
        best_hyperparams.append({
            'dataset': ds_name,
            'model': 'MLP',
            'hyperparams': f"hidden_size={mlp_params[0]}, lr={mlp_params[1]}"
        })
        
        # KMeans
        print("Tuning KMeans...")
        km_scores, km_params = tune_kmeans(X_tv, y_tv, X_test, y_test)
        all_results.append({
            'dataset': ds_name,
            'model': 'KMeans',
            'accuracy': km_scores['accuracy'],
            'f1_macro': km_scores['f1_macro'],
            'cohen_kappa': km_scores['cohen_kappa'],
            'silhouette': km_scores['silhouette'],
            'davies_bouldin': km_scores['davies_bouldin']
        })
        best_hyperparams.append({
            'dataset': ds_name,
            'model': 'KMeans',
            'hyperparams': f"n_clusters={km_params}"
        })
        
        # GMM
        print("Tuning GMM...")
        gmm_scores, gmm_params = tune_gmm(X_tv, y_tv, X_test, y_test)
        all_results.append({
            'dataset': ds_name,
            'model': 'GMM',
            'accuracy': gmm_scores['accuracy'],
            'f1_macro': gmm_scores['f1_macro'],
            'cohen_kappa': gmm_scores['cohen_kappa'],
            'silhouette': gmm_scores['silhouette'],
            'davies_bouldin': gmm_scores['davies_bouldin']
        })
        best_hyperparams.append({
            'dataset': ds_name,
            'model': 'GMM',
            'hyperparams': f"n_components={gmm_params}"
        })
        
        # SOM
        print("Tuning SOM...")
        som_scores, som_params = tune_som(X_tv, y_tv, X_test, y_test)
        all_results.append({
            'dataset': ds_name,
            'model': 'SOM',
            'accuracy': som_scores['accuracy'],
            'f1_macro': som_scores['f1_macro'],
            'cohen_kappa': som_scores['cohen_kappa'],
            'silhouette': som_scores['silhouette'],
            'davies_bouldin': som_scores['davies_bouldin']
        })
        best_hyperparams.append({
            'dataset': ds_name,
            'model': 'SOM',
            'hyperparams': f"grid_size={som_params[0]}, eta0={som_params[1]}"
        })
        
        print(f"\nCompleted dataset: {ds_name}")
        
    except Exception as e:
        print(f"\nError for dataset {ds_name}: {str(e)}")
        continue


df_new_results = pd.DataFrame(all_results)
df_new_hyperparams = pd.DataFrame(best_hyperparams)

# Merge with exsting results if file exists (avoid overwriting, except duplicates)
results_path = 'final_results_with_tuning.csv'
hyperparams_path = 'best_hyperparameters.csv'

if os.path.isfile(results_path):
    df_old_results = pd.read_csv(results_path)
    df_results = pd.concat([df_old_results, df_new_results], ignore_index=True)
    df_results = df_results.drop_duplicates(subset=['dataset', 'model'], keep='last')
else:
    df_results = df_new_results

if os.path.isfile(hyperparams_path):
    df_old_hyperparams = pd.read_csv(hyperparams_path)
    df_hyperparams = pd.concat([df_old_hyperparams, df_new_hyperparams], ignore_index=True)
    df_hyperparams = df_hyperparams.drop_duplicates(subset=['dataset', 'model'], keep='last')
else:
    df_hyperparams = df_new_hyperparams

# Save
df_results.to_csv(results_path, index=False)
df_hyperparams.to_csv(hyperparams_path, index=False)

# summary statistics
print("Perf by model:")
summary = df_results.groupby('model')[['accuracy', 'f1_macro', 'cohen_kappa']].mean()
print(summary.round(4))

## 6. Experimental Results

(The results can be seen below)

Key observations:
- Best average (accuracy): `RandomForest`.
- Best average (f1_macro): `RandomForest`.
- Several datasets show a large gap between `accuracy` and `f1_macro`, which signals class imbalance or majority-class predictions.  
  

### Dataset Evaluation

1. **Wine Dataset:**  
   - All models performed exceptionally well. Logistic Regression, Random Forest, and MLP achieved near-perfect scores.  
   - Clustering models (KMeans and GMM) also performed very well, with KMeans achieving perfect accuracy and a low Davies-Bouldin index, indicating well-separated clusters.

2. **Waveform Dataset:**  
   - Supervised models performed reasonably, with Logistic Regression and MLP achieving around 85% accuracy.  
   - Unsupervised models struggled: KMeans and GMM had lower accuracy and moderate silhouette scores, suggesting overlapping clusters. SOM performed poorly with negative silhouette scores.

3. **Car Dataset:**  
   - Random Forest excelled with perfect scores. Logistic Regression and MLP performed moderately.  
   - Clustering models struggled significantly, with KMeans and GMM accuracy around 70% and very low F1 scores, indicating poor separability in the feature space.

4. **Other Datasets (Summary Observations):**  
   - **High-performing datasets:** Wine, acute-inflammation, acute-nephritis, iris, and synthetic-control showed strong separability, leading to high performance across models.  
   - **Challenging datasets:** Zoo, tic-tac-toe, balance-scale, statlog-vehicle, and waveform had lower clustering performance and moderate supervised learning accuracy, reflecting either class imbalance or complex feature interactions.

5. **Unsupervised Evaluation:**  
   - **Silhouette Score:** High values (>0.4) indicate compact and well-separated clusters, observed in Wine, Acute-inflammation, and Iris datasets.  
   - **Davies-Bouldin Index:** Low values (<1.5) indicate better clustering; datasets such as acute-inflammation and wine show well-defined clusters.  
   - **SOM Performance:** SOM often underperforms relative to KMeans or GMM, showing negative silhouette scores in many datasets, suggesting less effective clustering in high-dimensional or complex datasets.


### Discussion

- Supervised models (Random Forest, Logistic Regression, MLP) generally outperform unsupervised models in datasets with well-defined class labels.  
- Clustering algorithms are effective only on datasets with clear cluster separation (e.g., wine, acute-inflammation, iris).  
- SOM is less reliable for clustering in datasets with high noise or overlapping classes.  
- Dataset characteristics (number of classes, feature dimensionality, and separability) significantly impact model performance.



In [26]:
df = pd.read_csv("final_results_with_tuning.csv")
for col in df.columns:
    df[col] = pd.to_numeric(df[col], errors="ignore")

# mean scores by method
metrics = ["accuracy", "f1_macro", "cohen_kappa", "silhouette", "davies_bouldin"]
mean_df = (
    df.groupby("model")[metrics]
    .mean()
    .reset_index()
)
md = mean_df.to_markdown(index=False, floatfmt=".3f", tablefmt="pipe")
print(md)

# details
df = df.sort_values(["dataset", "model"])
df = df.fillna("–")
md = df.to_markdown(index=False, floatfmt=".3f")
display(Markdown(md))


| model              |   accuracy |   f1_macro |   cohen_kappa |   silhouette |   davies_bouldin |
|:-------------------|-----------:|-----------:|--------------:|-------------:|-----------------:|
| GMM                |      0.782 |      0.654 |         0.568 |        0.264 |            1.758 |
| KMeans             |      0.754 |      0.627 |         0.536 |        0.313 |            1.414 |
| LogisticRegression |      0.907 |      0.876 |         0.818 |      nan     |          nan     |
| MLP                |      0.870 |      0.800 |         0.764 |      nan     |          nan     |
| RandomForest       |      0.904 |      0.879 |         0.817 |      nan     |          nan     |
| SOM                |      0.718 |      0.591 |         0.464 |       -0.059 |            2.137 |


C:\Users\blue416yt\AppData\Local\Temp\ipykernel_15716\1503484408.py:3: FutureWarning: errors='ignore' is deprecated and will raise in a future version. Use to_numeric without passing `errors` and catch exceptions explicitly instead
  df[col] = pd.to_numeric(df[col], errors="ignore")


| dataset                  | model              |   accuracy |   f1_macro |   cohen_kappa | silhouette          | davies_bouldin     |
|:-------------------------|:-------------------|-----------:|-----------:|--------------:|:--------------------|:-------------------|
| acute-inflammation       | GMM                |      1.000 |      1.000 |         1.000 | 0.5957184339984067  | 0.6854540573738328 |
| acute-inflammation       | KMeans             |      0.958 |      0.958 |         0.917 | 0.4525143296552947  | 1.1293800865263803 |
| acute-inflammation       | LogisticRegression |      1.000 |      1.000 |         1.000 | –                   | –                  |
| acute-inflammation       | MLP                |      1.000 |      1.000 |         1.000 | –                   | –                  |
| acute-inflammation       | RandomForest       |      1.000 |      1.000 |         1.000 | –                   | –                  |
| acute-inflammation       | SOM                |      0.583 |      0.571 |         0.167 | 0.1667966912008561  | 0.9605350720711244 |
| acute-nephritis          | GMM                |      1.000 |      1.000 |         1.000 | 0.5691474816019677  | 0.7417975036175986 |
| acute-nephritis          | KMeans             |      1.000 |      1.000 |         1.000 | 0.5691474816019677  | 0.7417975036175986 |
| acute-nephritis          | LogisticRegression |      1.000 |      1.000 |         1.000 | –                   | –                  |
| acute-nephritis          | MLP                |      1.000 |      1.000 |         1.000 | –                   | –                  |
| acute-nephritis          | RandomForest       |      1.000 |      1.000 |         1.000 | –                   | –                  |
| acute-nephritis          | SOM                |      1.000 |      1.000 |         1.000 | 0.5691474816019677  | 0.7417975036175986 |
| balance-scale            | GMM                |      0.632 |      0.434 |         0.317 | 0.179885365686974   | 1.4386692825779197 |
| balance-scale            | KMeans             |      0.600 |      0.400 |         0.256 | 0.1807224797941536  | 1.6652050849417932 |
| balance-scale            | LogisticRegression |      0.872 |      0.696 |         0.770 | –                   | –                  |
| balance-scale            | MLP                |      0.872 |      0.606 |         0.763 | –                   | –                  |
| balance-scale            | RandomForest       |      0.784 |      0.578 |         0.629 | –                   | –                  |
| balance-scale            | SOM                |      0.536 |      0.374 |         0.146 | -0.319873020952549  | 3.442415362559188  |
| car                      | GMM                |      0.699 |      0.206 |         0.000 | 0.1203655651650388  | 2.029697544332389  |
| car                      | KMeans             |      0.699 |      0.206 |         0.000 | 0.1280805247308292  | 2.0150295805980667 |
| car                      | LogisticRegression |      0.818 |      0.707 |         0.606 | –                   | –                  |
| car                      | MLP                |      0.841 |      0.635 |         0.639 | –                   | –                  |
| car                      | RandomForest       |      1.000 |      1.000 |         1.000 | –                   | –                  |
| car                      | SOM                |      0.694 |      0.212 |         0.015 | -0.2875469920406069 | 3.232206609960194  |
| dermatology              | GMM                |      0.757 |      0.550 |         0.690 | 0.2087772880174884  | 2.186478774968842  |
| dermatology              | KMeans             |      0.716 |      0.583 |         0.641 | 0.3037914333631948  | 1.2929939411856444 |
| dermatology              | LogisticRegression |      0.973 |      0.970 |         0.966 | –                   | –                  |
| dermatology              | MLP                |      0.946 |      0.942 |         0.932 | –                   | –                  |
| dermatology              | RandomForest       |      0.973 |      0.969 |         0.966 | –                   | –                  |
| dermatology              | SOM                |      0.770 |      0.623 |         0.708 | -0.026834961973195  | 1.8518897328821584 |
| ecoli                    | GMM                |      0.750 |      0.498 |         0.638 | 0.4009945060927541  | 0.7362364866722875 |
| ecoli                    | KMeans             |      0.765 |      0.403 |         0.661 | 0.379155477367912   | 0.8982452275424112 |
| ecoli                    | LogisticRegression |      0.838 |      0.780 |         0.775 | –                   | –                  |
| ecoli                    | MLP                |      0.794 |      0.512 |         0.714 | –                   | –                  |
| ecoli                    | RandomForest       |      0.868 |      0.872 |         0.816 | –                   | –                  |
| ecoli                    | SOM                |      0.750 |      0.390 |         0.632 | -0.2008590808086184 | 1.464339358710935  |
| heart-hungarian          | GMM                |      0.814 |      0.794 |         0.589 | 0.1901241487568283  | 2.1632731948247867 |
| heart-hungarian          | KMeans             |      0.847 |      0.828 |         0.656 | 0.182195783022142   | 2.03655250724934   |
| heart-hungarian          | LogisticRegression |      0.831 |      0.806 |         0.614 | –                   | –                  |
| heart-hungarian          | MLP                |      0.797 |      0.760 |         0.526 | –                   | –                  |
| heart-hungarian          | RandomForest       |      0.797 |      0.767 |         0.537 | –                   | –                  |
| heart-hungarian          | SOM                |      0.864 |      0.852 |         0.704 | -0.1127232920401541 | 1.4972901089620678 |
| ionosphere               | SOM                |      0.845 |      0.834 |         0.669 | -0.150151488504233  | 1.7469284995834509 |
| iris                     | GMM                |      1.000 |      1.000 |         1.000 | 0.3879555851415599  | 1.0386632303083412 |
| iris                     | KMeans             |      0.800 |      0.800 |         0.700 | 0.4851298055402769  | 0.7968303222944372 |
| iris                     | LogisticRegression |      1.000 |      1.000 |         1.000 | –                   | –                  |
| iris                     | MLP                |      0.867 |      0.865 |         0.800 | –                   | –                  |
| iris                     | RandomForest       |      0.933 |      0.933 |         0.900 | –                   | –                  |
| iris                     | SOM                |      0.700 |      0.616 |         0.550 | 0.1083112811697075  | 1.4721184871977504 |
| page-blocks              | GMM                |      0.898 |      0.189 |         0.000 | 0.1726100477545608  | 2.0481092402491408 |
| page-blocks              | KMeans             |      0.919 |      0.370 |         0.344 | 0.5107886938501736  | 0.945587653900734  |
| page-blocks              | LogisticRegression |      0.965 |      0.849 |         0.810 | –                   | –                  |
| page-blocks              | MLP                |      0.966 |      0.816 |         0.810 | –                   | –                  |
| page-blocks              | RandomForest       |      0.978 |      0.869 |         0.881 | –                   | –                  |
| page-blocks              | SOM                |      0.898 |      0.189 |         0.000 | -0.2828177327149508 | 3.1984650089082414 |
| seeds                    | GMM                |      0.857 |      0.851 |         0.786 | 0.3136610744106988  | 1.0674399126060736 |
| seeds                    | KMeans             |      0.810 |      0.801 |         0.714 | 0.3468328475781327  | 1.0920077909017063 |
| seeds                    | LogisticRegression |      0.905 |      0.901 |         0.857 | –                   | –                  |
| seeds                    | MLP                |      0.833 |      0.828 |         0.750 | –                   | –                  |
| seeds                    | RandomForest       |      0.905 |      0.901 |         0.857 | –                   | –                  |
| seeds                    | SOM                |      0.786 |      0.751 |         0.679 | -0.0712347168366568 | 1.4674391605412085 |
| spambase                 | GMM                |      0.792 |      0.774 |         0.550 | -0.0340210681032222 | 5.733542003389461  |
| spambase                 | KMeans             |      0.831 |      0.811 |         0.627 | 0.1266556921782768  | 2.551296278020045  |
| spambase                 | LogisticRegression |      0.926 |      0.922 |         0.845 | –                   | –                  |
| spambase                 | MLP                |      0.929 |      0.926 |         0.852 | –                   | –                  |
| spambase                 | RandomForest       |      0.934 |      0.930 |         0.859 | –                   | –                  |
| spambase                 | SOM                |      0.821 |      0.806 |         0.614 | -0.3348732106932315 | 3.904749864215471  |
| statlog-heart            | GMM                |      0.759 |      0.759 |         0.522 | 0.1406439999937895  | 2.3966635490860626 |
| statlog-heart            | KMeans             |      0.815 |      0.812 |         0.625 | 0.1672129284930099  | 2.1531169253105595 |
| statlog-heart            | LogisticRegression |      0.852 |      0.850 |         0.700 | –                   | –                  |
| statlog-heart            | MLP                |      0.852 |      0.852 |         0.705 | –                   | –                  |
| statlog-heart            | RandomForest       |      0.833 |      0.832 |         0.664 | –                   | –                  |
| statlog-heart            | SOM                |      0.815 |      0.811 |         0.622 | -0.1985843574099964 | 1.3941278430120394 |
| statlog-vehicle          | GMM                |      0.424 |      0.432 |         0.230 | 0.1918910373806462  | 1.8181914405871    |
| statlog-vehicle          | KMeans             |      0.371 |      0.246 |         0.173 | 0.3039280610461231  | 1.0814434234627524 |
| statlog-vehicle          | LogisticRegression |      0.835 |      0.832 |         0.780 | –                   | –                  |
| statlog-vehicle          | MLP                |      0.800 |      0.793 |         0.733 | –                   | –                  |
| statlog-vehicle          | RandomForest       |      0.759 |      0.751 |         0.678 | –                   | –                  |
| statlog-vehicle          | SOM                |      0.359 |      0.291 |         0.152 | -0.1634849488190448 | 2.1278531023948903 |
| synthetic-control        | GMM                |      0.483 |      0.322 |         0.380 | 0.2839710966708134  | 1.47216972408268   |
| synthetic-control        | KMeans             |      0.608 |      0.492 |         0.530 | 0.2616590646844599  | 1.5001437440522143 |
| synthetic-control        | LogisticRegression |      0.975 |      0.974 |         0.970 | –                   | –                  |
| synthetic-control        | MLP                |      0.992 |      0.992 |         0.990 | –                   | –                  |
| synthetic-control        | RandomForest       |      0.983 |      0.983 |         0.980 | –                   | –                  |
| synthetic-control        | SOM                |      0.642 |      0.530 |         0.570 | 0.0477274000390488  | 2.833409691557174  |
| tic-tac-toe              | GMM                |      0.729 |      0.669 |         0.350 | 0.0505793340148437  | 2.345918253742451  |
| tic-tac-toe              | KMeans             |      0.651 |      0.580 |         0.172 | 0.1106852124410439  | 2.197351313182001  |
| tic-tac-toe              | LogisticRegression |      0.974 |      0.971 |         0.942 | –                   | –                  |
| tic-tac-toe              | MLP                |      0.974 |      0.971 |         0.942 | –                   | –                  |
| tic-tac-toe              | RandomForest       |      0.979 |      0.977 |         0.954 | –                   | –                  |
| tic-tac-toe              | SOM                |      0.599 |      0.547 |         0.095 | -0.2009817909697087 | 3.501987433473247  |
| titanic                  | GMM                |      0.757 |      0.699 |         0.404 | 0.6619438091591356  | 0.5485973246313095 |
| titanic                  | KMeans             |      0.757 |      0.699 |         0.404 | 0.6619438091591356  | 0.5485973246313095 |
| titanic                  | LogisticRegression |      0.766 |      0.700 |         0.409 | –                   | –                  |
| titanic                  | MLP                |      0.766 |      0.700 |         0.409 | –                   | –                  |
| titanic                  | RandomForest       |      0.778 |      0.679 |         0.392 | –                   | –                  |
| titanic                  | SOM                |      0.753 |      0.705 |         0.411 | 0.4065991846515469  | 1.042952273176141  |
| vertebral-column-2clases | GMM                |      0.790 |      0.777 |         0.561 | 0.1373194225360157  | 2.169823055372861  |
| vertebral-column-2clases | KMeans             |      0.629 |      0.621 |         0.275 | 0.2971303009398017  | 1.0130104615230358 |
| vertebral-column-2clases | LogisticRegression |      0.839 |      0.815 |         0.631 | –                   | –                  |
| vertebral-column-2clases | MLP                |      0.855 |      0.827 |         0.654 | –                   | –                  |
| vertebral-column-2clases | RandomForest       |      0.790 |      0.763 |         0.526 | –                   | –                  |
| vertebral-column-2clases | SOM                |      0.629 |      0.628 |         0.320 | -0.1668337449014195 | 2.097118135117669  |
| vertebral-column-3clases | GMM                |      0.774 |      0.565 |         0.622 | 0.2645073046624385  | 1.3589261704843674 |
| vertebral-column-3clases | KMeans             |      0.613 |      0.450 |         0.360 | 0.3101397254305879  | 1.249539329019484  |
| vertebral-column-3clases | LogisticRegression |      0.855 |      0.804 |         0.766 | –                   | –                  |
| vertebral-column-3clases | MLP                |      0.758 |      0.716 |         0.623 | –                   | –                  |
| vertebral-column-3clases | RandomForest       |      0.855 |      0.815 |         0.766 | –                   | –                  |
| vertebral-column-3clases | SOM                |      0.613 |      0.529 |         0.391 | -0.130030125882062  | 1.796103793343736  |
| waveform                 | GMM                |      0.767 |      0.755 |         0.650 | 0.1126398900235161  | 2.131259454024782  |
| waveform                 | KMeans             |      0.638 |      0.615 |         0.455 | 0.165394482501911   | 1.8531360146657128 |
| waveform                 | LogisticRegression |      0.855 |      0.854 |         0.782 | –                   | –                  |
| waveform                 | MLP                |      0.854 |      0.853 |         0.781 | –                   | –                  |
| waveform                 | RandomForest       |      0.836 |      0.835 |         0.754 | –                   | –                  |
| waveform                 | SOM                |      0.676 |      0.617 |         0.513 | -0.0750284344757483 | 4.825065427894199  |
| wine                     | GMM                |      0.972 |      0.972 |         0.958 | 0.2829173375199325  | 1.3867963486679635 |
| wine                     | KMeans             |      1.000 |      1.000 |         1.000 | 0.2493705353772339  | 1.728805413157303  |
| wine                     | LogisticRegression |      0.972 |      0.971 |         0.958 | –                   | –                  |
| wine                     | MLP                |      1.000 |      1.000 |         1.000 | –                   | –                  |
| wine                     | RandomForest       |      1.000 |      1.000 |         1.000 | –                   | –                  |
| wine                     | SOM                |      0.750 |      0.743 |         0.630 | -0.0049090577402939 | 1.2270550614055749 |
| zoo                      | GMM                |      0.762 |      0.484 |         0.683 | 0.3179226831140359  | 1.4111895863890147 |
| zoo                      | KMeans             |      0.810 |      0.490 |         0.752 | 0.3729852992079202  | 1.2009813182213434 |
| zoo                      | LogisticRegression |      1.000 |      1.000 |         1.000 | –                   | –                  |
| zoo                      | MLP                |      0.571 |      0.206 |         0.411 | –                   | –                  |
| zoo                      | RandomForest       |      1.000 |      1.000 |         1.000 | –                   | –                  |
| zoo                      | SOM                |      0.714 |      0.374 |         0.619 | 0.1394541860069769  | 1.1908240943441224 |

### Comparison with baseline performance reported in the paper

The relevant baseline results from the paper are summarized below.

**Logistic Regression (LMR)**

Average accuracy: 76.5%

Cohen’s kappa: 53.7%

Friedman ranking: 78.5

**Random Forest (RF)**

Average accuracy: 80.0%

Cohen’s kappa: 57.4%

Friedman ranking: 60.5

**Multilayer Perceptron (NNET)**

Average accuracy: 80.3%

Cohen’s kappa: 62.0%

Friedman ranking: 47.3  
  
  
#### Comparison with our results :  
  
**Logistic Regression**

Our accuracy: 90.7%

Our Cohen’s kappa: 81.8%

Compared to the baseline (76.5% accuracy, 53.7% kappa), our results show a substantial improvement. This difference can be explained by the use of a smaller, selected subset of datasets and different preprocessing and evaluation protocols.

**Random Forest**

Our accuracy: 90.4%

Our Cohen’s kappa: 81.7%

These results clearly outperform the baseline values reported in the paper (80.0% accuracy, 57.4% kappa).

**Multilayer Perceptron (MLP)**

Our accuracy: 87.0%

Our Cohen’s kappa: 76.4%

Although slightly below Logistic Regression and Random Forest in our experiments (although its Friedman ranking in the paper identifies it as the best-performing model of these three), the model still exceeds the average baseline performance reported in the paper (80.3% accuracy, 62.0% kappa).


#### Discussion

The supervised models evaluated in our experiments match or exceed the average baseline performance reported in the paper.

Absolute performance differences are expected, as the paper reports results averaged over 121 datasets, whereas our experiments use a smaller, curated subset. Also, I admit that I deliberately avoided the biggest and most complex datasets, which may have played a role in this result.

Unsupervised models (GMM, K-Means, SOM) are not directly comparable to the baselines in the paper as it focuses on supervised classification.

## 7. Conclusions  

From our experiments, we can see that supervised models like Logistic Regression, Random Forest, and MLP work very good on many datasets. When we compare with the baseline in the paper, our models have better results in both accuracy and Cohen's kappa. Logistic Regression and Random Forest have more than 90% accuracy and more than 81% kappa, and MLP also is better than baseline with 87% accuracy and 76.4% kappa.

The reason our results are higher maybe because:
- We use smaller subset of datasets, not all 121 datasets.  
- Different preprocessing and evaluation maybe help the models.  
- I may have avoid some of the biggest and hard datasets, so performance is easier.

Unsupervised models (K-Means, GMM, SOM) are not same as supervised, so we cannot really compare with the paper. They are designed for finding patterns or clusters in data, not for predicting known labels. They work better if datasets have clear clusters. SOM often do worse, especially in many features or overlapping clusters.

Some important points:
- Supervised models usually work well for classification tasks.  
- Clustering works better on simple and clean datasets with clear structure.  
- Choosing dataset and preprocessing carefully can make big difference for results.

In general, our experiments show standard supervised models can get strong results if used carefully, and unsupervised methods give useful information about dataset structure, but they are not meant to be compared on classification performance.